src/Microsoft.OData.Edm/Schema/EdmEntityType.cs

```diff
@@ -157,6 +158,37 @@ public void AddKeys(IEnumerable<IEdmStructuralProperty> keyProperties)
                     this.declaredKey = new List<IEdmStructuralProperty>();
                 }
 
+                //The Key property is the last segment in the path segment. 
+                IEdmStructuralPropertyAlias structuralPropertyAlias = property as IEdmStructuralPropertyAlias;
+                if (!string.IsNullOrEmpty(structuralPropertyAlias?.PropertyAlias))
+                {
+                    string[] keyPropertySegments = structuralPropertyAlias.Path.ToArray();
+                    IEdmStructuralProperty structuralProperty = this.FindProperty(keyPropertySegments[0]) as IEdmStructuralProperty;
+                    for (int i = 1; i < keyPropertySegments.Length; i++)
+                    {
+                        if (structuralProperty.Type.Definition.TypeKind == EdmTypeKind.Complex)
+                        {
+                            structuralProperty = structuralProperty.Type.AsComplex().FindProperty(keyPropertySegments[i]) as IEdmStructuralProperty;
+                        }
+                        else if (structuralProperty.Type.Definition.TypeKind == EdmTypeKind.Entity)
+                        {
+                            structuralProperty = structuralProperty.Type.AsEntity().FindProperty(keyPropertySegments[i]) as IEdmStructuralProperty;
+                        }
+                        else
+                        {
+                            throw new InvalidOperationException(Strings.Bad_UnresolvedType(structuralProperty));
+                        }
+                    }
+
+                    if (structuralProperty != null)
+                    {
+                        structuralProperty = new EdmStructuralPropertyAlias(structuralProperty.DeclaringType, structuralProperty.Name, structuralProperty.Type, structuralPropertyAlias.PropertyAlias, keyPropertySegments);

```

Comment:
Why create a new `EdmStructuralPropertyAlias` if the `property` is already an `IEdmStructuralPropertyAlias`

Reply:
The `property` is not the key property - This is the complex property - that is why we need to loop through the path segments to get to the last path segment which will be the desired key property and create an `EdmStructuralPropertyAlias` for it.